pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,726 @@
+                ensure!(amount > current_highest_bid.1, Error::<T>::InvalidBidAmount);
+            }
+
+            // reject if user already has a bid
+            ensure!(
+                auction.bidders.find_bid(who.clone()).is_none(),
+                Error::<T>::UserBidAlreadyExists
+            );
+
+            // transfer funds from caller
+            T::Currency::transfer(&who, &Self::account_id(), amount, KeepAlive)?;
+
+            // add bid to storage
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                // should not panic when unwrap since already checked above
+                let auction = maybe_auction
+                    .as_mut()
+                    .ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+                // extend auction by grace period if in ending period
+                if auction.end_block.checked_sub(&current_block)
+                    <= Some(T::AuctionEndingPeriod::get())
+                {
+                    auction.end_block = auction
+                        .end_block
+                        .checked_add(&T::AuctionGracePeriod::get())
+                        .ok_or(Error::<T>::UnexpectedError)?;
+                }
+
+                auction.state = AuctionState::Extended;
+
+                // replace top bidder with caller
+                // if bidder has been removed, refund removed user
+                if let Some(bid) = auction.bidders.insert_new_bid(who.clone(), amount) {
+                    T::Currency::transfer(&Self::account_id(), &bid.0, bid.1, KeepAlive)?;
+                }
+
+                Ok(())
+            })?;
+
+            // emit bid created event
+            Self::deposit_event(Event::BidCreated {
+                nft_id,
+                marketplace_id: auction.marketplace_id,
+                creator: who,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn remove_bid(origin: OriginFor<T>, nft_id: NFTId) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+            // remove bid from storage
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                // should not panic when unwrap since already checked above
+                let auction = maybe_auction
+                    .as_mut()
+                    .ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+                // ensure the auction period has not ended
+                ensure!(auction.end_block > current_block, Error::<T>::AuctionEnded);
+
+                let user_bid = auction
+                    .bidders
+                    .remove_bid(who.clone())
+                    .ok_or(Error::<T>::BidDoesNotExist)?;
+
+                T::Currency::transfer(&Self::account_id(), &who, user_bid.1, KeepAlive)?;
```

Comment:
It should be AllowDeath instead of KeepAlive. It's pretty normal for the pallet account to have 0 CAPS. If we are using KeepAlive, the system will not allow us to move all the funds from the pallet's account.